drivers/crypto/crypto_smartbond.c

```diff
@@ -0,0 +1,1049 @@
+		return -EINVAL;
+	}
+
+	AES_HASH->CRYPTO_MREG0_REG = 0;
+	AES_HASH->CRYPTO_MREG1_REG = 0;
+	AES_HASH->CRYPTO_MREG2_REG = 0;
+	AES_HASH->CRYPTO_MREG3_REG = 0;
+
+	crypto_smartbond_cipher_store_dep_data((uint32_t *)iv, len_words);
+
+	/* Return success */
+	return 0;
+}
+
+static int crypto_smartbond_cipher_store_ic(uint8_t *ic, uint32_t len_words)
+{
+	if (ic == NULL || len_words == 0 || len_words > 4) {
+		return -EINVAL;
+	}
+
+	AES_HASH->CRYPTO_MREG0_REG = 0;
+	AES_HASH->CRYPTO_MREG1_REG = 0;
+	AES_HASH->CRYPTO_MREG2_REG = 0;
+	AES_HASH->CRYPTO_MREG3_REG = 0;
+
+	crypto_smartbond_cipher_store_dep_data((uint32_t *)ic, len_words);
+
+	/* Return success */
+	return 0;
+}
+
+static void crypto_smartbond_set_status(bool enable)
+{
+	unsigned int key;
+
+	key = irq_lock();
+
+	if (enable) {
+		CRG_TOP->CLK_AMBA_REG |= (CRG_TOP_CLK_AMBA_REG_AES_CLK_ENABLE_Msk);
+
+		AES_HASH->CRYPTO_CLRIRQ_REG = 0x1;
+		AES_HASH->CRYPTO_CTRL_REG |= (AES_HASH_CRYPTO_CTRL_REG_CRYPTO_IRQ_EN_Msk);
+
+		NVIC_ClearPendingIRQ(IRQN);
+		NVIC_EnableIRQ(IRQN);
+	} else {
+		AES_HASH->CRYPTO_CTRL_REG &= ~(AES_HASH_CRYPTO_CTRL_REG_CRYPTO_IRQ_EN_Msk);
+		AES_HASH->CRYPTO_CLRIRQ_REG = 0x1;
+
+		NVIC_DisableIRQ(IRQN);
+		NVIC_ClearPendingIRQ(IRQN);
+
+		CRG_TOP->CLK_AMBA_REG &= ~(CRG_TOP_CLK_AMBA_REG_AES_CLK_ENABLE_Msk);
+	}
+
+	irq_unlock(key);
+}
+
+static int crypto_smartbond_query_hw_caps(const struct device *dev)
+{
+	return CRYPTO_HW_CAPS;
+}
+
+static int crypto_smartbond_cipher_ecb_handler(struct cipher_ctx *ctx, struct cipher_pkt *pkt)
+{
+	int ret;
+	struct crypto_smartbond_data *data = ctx->device->data;
+
+	if ((AES_HASH->CRYPTO_STATUS_REG & AES_HASH_CRYPTO_STATUS_REG_CRYPTO_INACTIVE_Msk) == 0) {
+		LOG_ERR("Crypto engine is already employed");
+		return -EINVAL;
+	}
+
+	if (pkt->out_buf_max < pkt->in_len) {
+		LOG_ERR("OUT buffer cannot be less that IN buffer");
+		return -EINVAL;
+	}
+
+	if (pkt->in_buf == NULL || pkt->out_buf == NULL) {
+		LOG_ERR("Missing IN or OUT buffer declaration");
+		return -EIO;
+	}
+
+	if (pkt->in_len > 16) {
+		LOG_ERR("For security reasons, do not operate on more than 16 bytes");
+		return -EINVAL;
+	}
+
+	k_sem_take(&data->device_sem, K_FOREVER);
+
+	ret = crypto_smartbond_check_in_restrictions(pkt->in_len);
+	if (ret < 0) {
+		LOG_ERR("Unsupported IN buffer size");
+		k_sem_give(&data->device_sem);
+		return ret;
+	}
+
+	ret = crypto_smartbond_set_in_out_buf(pkt->in_buf, pkt->out_buf, pkt->in_len);
+	if (ret < 0) {
+		LOG_ERR("Unsupported IN or OUT buffer location");
+		k_sem_give(&data->device_sem);
+		return ret;
+	}
+
+#if defined(CONFIG_CRYPTO_ASYNC)
+	data->cipher_pkt = pkt;
+#endif
+
+	/* Start crypto processing */
+	AES_HASH->CRYPTO_START_REG = 1;
+
+#if !defined(CONFIG_CRYPTO_ASYNC)
+	/* Wait for crypto to finish its task */
+	k_sem_take(&data->sync_sem, K_FOREVER);
+#endif
+
+	/* Report that number of bytes operated upon. */
+	pkt->out_len = pkt->in_len;
+
+	k_sem_give(&data->device_sem);
```

Comment:
how does this work in async mode if you release device sem before operation is finished?

Reply:
In async. mode, user has to register a callback function which is called when the current cryptographic operation is complete. It's the responsibility of the developer to make sure that no other cryptographic operations take place as long as the callback is fired, designating the completion of the previous cryptographic operations. This is the benefit of performing async. operations isn't it?  All of the cryptographic operations will first check if crypto is in progress and will return with error code. -EINVAL.